src/AudioCaptureImpl_WASAPI.cpp

```diff
@@ -1,25 +1,631 @@
+        if (FAILED(result))
+        {
+            poco_error_f1(_logger, "IMMDeviceEnumerator::EnumAudioEndpoints failed: result = 0x%08?x", result);
+            return;
+        }
+
+        UINT deviceCount{0};
+        audioEndpoints->GetCount(&deviceCount);
+        for (UINT item = 0; item < deviceCount; item++)
+        {
+            IMMDevice* device{nullptr};
+            result = audioEndpoints->Item(item, &device);
+
+            if (FAILED(result) || device == nullptr)
+            {
+                poco_error_f2(_logger, "IMMDeviceEnumerator::Item failed for device %?u: result = 0x%08?x", item, result);
+                continue;
+            }
+
+            deviceList.emplace_back(device, (dataFlow == eRender));
+            device->Release();
+        }
+
+        audioEndpoints->Release();
+    };
+
+    std::vector<AudioDevice> deviceList;
+    addEndpoints(deviceList, eRender);
+    addEndpoints(deviceList, eCapture);
+
+    return deviceList;
+}
+
+bool AudioCaptureImpl::OpenAudioDevice(IMMDevice* device, bool useLoopback)
+{
+    // activate an IAudioClient
+    HRESULT result = device->Activate(__uuidof(IAudioClient),
+                              CLSCTX_ALL,
+                              nullptr,
+                              reinterpret_cast<void**>(&_audioClient));
+    if (FAILED(result))
+    {
+        poco_error_f1(_logger, "IMMDevice::Activate(IAudioClient) failed: result = 0x%08?x", result);
+        return false;
+    }
+
+    // get the default device periodicity
+    REFERENCE_TIME hnsDefaultDevicePeriod;
+    result = _audioClient->GetDevicePeriod(&hnsDefaultDevicePeriod,
+                                           nullptr);
+    if (FAILED(result))
+    {
+        poco_error_f1(_logger, "IAudioClient::GetDevicePeriod failed: result = 0x%08?x", result);
+        return false;
+    }
+
+    // get the default device format
+    WAVEFORMATEX* pwfx;
+    result = _audioClient->GetMixFormat(&pwfx);
+    if (FAILED(result))
+    {
+        poco_error_f1(_logger, "IAudioClient::GetMixFormat failed: result = 0x%08?x", result);
+        return false;
+    }
+
+    // Should default to float32 data, but some devices might deliver other formats.
+    if (pwfx->wFormatTag != WAVE_FORMAT_IEEE_FLOAT)
+    {
+        auto extensibleFormat = reinterpret_cast<PWAVEFORMATEXTENSIBLE>(pwfx);
+        if (pwfx->wFormatTag != WAVE_FORMAT_EXTENSIBLE)
+        {
+            poco_error_f1(_logger, "IAudioClient::GetMixFormat returned non-float sample format: 0x%04?x", pwfx->wFormatTag);
+            CoTaskMemFree(pwfx);
+            return false;
+        }
+        else if (!IsEqualGUID(KSDATAFORMAT_SUBTYPE_IEEE_FLOAT, extensibleFormat->SubFormat))
+        {
+            poco_error_f1(_logger, "IAudioClient::GetMixFormat returned non-float extensible sub format: 0x%04?x", extensibleFormat->SubFormat);
+            CoTaskMemFree(pwfx);
+            return false;
+        }
+    }
+
+    _channels = pwfx->nChannels;
```

Comment:
This is actually important, as loopback capture devices will use the same number of channels as the original output device. This means, on a device configured for 5.1 output, we get 6 channels/samples per frame.

It's not guaranteed that front left/right channels are the first two. We could cast `pwfx` to `WAVEFORMATEXTENSIBLE` and then retrieve the channel-to-speaker assignment, but since projectM doesn't support providing channel indices, we'd then need to repack the buffer data, similar to what was done in the Apple Music plug-in.

Reply:
What's a scenario where the first two channels are not L/R?

Reply:
We just need to pass the correct channel number to projectM to access the proper one or two samples inside each frame.